build-llvm.py

```diff
@@ -0,0 +1,402 @@
+    if not 'LD' in os.environ:
+        # and we're using clang, try to find the fastest one
+        if "clang" in cc:
+            possible_linkers = ['lld-9', 'lld-8', 'lld-7', 'lld', 'gold', 'bfd']
+            for linker in possible_linkers:
+                # We want to find lld wherever the clang we are using is located
+                ld = shutil.which("ld." + linker,
+                                  path=cc_folder + ":" + os.environ['PATH'])
+                if ld is not None:
+                    break
+            # If clang is older than 3.9, it won't accept absolute paths so we
+            # need to just pass it the name (and modify PATH so that it is found properly)
+            # https://github.com/llvm/llvm-project/commit/e43b7413597d8102a4412f9de41102e55f4f2ec9
+            if clang_version(cc) < 390:
+                os.environ['PATH'] = cc_folder + ":" + os.environ['PATH']
+                ld = linker
+        # and we're using gcc, try to use gold
+        else:
+            ld = "gold"
+            if linker_test(cc, ld):
+                ld = None
+    # If the user did specify a linker
+    else:
+        # evaluate its full path with clang to avoid weird issues and check to
+        # see if it will work with '-fuse-ld', which is what cmake will do. Doing
+        # it now prevents a hard error later.
+        ld = os.environ['LD']
+        if "clang" in cc and clang_version(cc) >= 390:
+            ld = shutil.which(ld)
+        if linker_test(cc, ld):
+            print("LD won't work with " + cc +
+                  ", saving you from yourself by ignoring LD value")
+            ld = None
+
+    # Print what binaries we are using to compile/link with so the user can
+    # decide if that is proper or not
+    print("CC: " + cc)
+    print("CXX: " + cxx)
+    if ld is not None:
+        ld = ld.rstrip()
+        ld_to_print = shutil.which("ld." + ld)
+        if ld_to_print is None:
+            ld_to_print = shutil.which(ld)
+        print("LD: " + ld_to_print)
+
+    return cc, cxx, ld
+
+
+# Make sure that the base dependencies of cmake, curl, git, and ninja are installed
+def check_dependencies():
+    utils.header("Checking dependencies")
+    required_commands = ["cmake", "curl", "git", "ninja"]
+    for command in required_commands:
+        output = shutil.which(command)
+        if output is None:
+            raise RuntimeError(command +
+                               " could not be found, please install it!")
+        print(output)
+
+
+# Download llvm and binutils or update them if they exist
+def fetch_llvm_binutils(root, update, branch):
+    p = root.joinpath("llvm-project")
+    if p.is_dir():
+        if update:
+            utils.header("Updating LLVM")
+            subprocess.run(
+                ["git", "-C", p.as_posix(), "checkout", branch], check=True)
+            subprocess.run(
+                ["git", "-C", p.as_posix(), "pull", "--rebase"], check=True)
```

Comment:
Hmmm, I suppose it won't (it won't work for a tag either).

```git ch v5.1-rc5 && git pull
Note: checking out 'v5.1-rc5'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at dc4060a5dc25 Linux 5.1-rc5
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>
```

At the same time, if someone wants to check out at a tag or a commit, they can do it manually in the repository then pass `-n` to the script to avoid updating. I am happy to hear suggestions for getting around that though.